### PR TITLE
Template workflows

### LINKED ISSUE #20: RF: so there is little to no duplication
ATM multiple steps across workflows across different OSs are duplicated.  Let's come up with some way to mint different workflows from a single specification.

### PR DESCRIPTION
**Pull request by jwodder**
Closes #20.

### CODE REVIEW

File: .github/workflows/template/template.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,537 @@
+name: Build git-annex on {{osname}}
```
Comment (yarikoptic) [inline .github/workflows/template/template.yml.j2:1]: name the file `build-{{osname}}.yml.j2` and use the same dictionary to "format"?

File: .github/workflows/build-windows.yaml [hunk 2]
```diff
@@ -196,114 +205,4 @@ jobs:
 
             See <https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}> for more information.
 
-  test-datalad:
```
Comment (yarikoptic) [inline .github/workflows/build-windows.yaml:199]: test-datalad I guess is yet to reincarnate, just making sure it is not lost

**Review by yarikoptic — COMMENTED** [review]
initial comments

File: .github/workflows/template/template.yml.j2 (hunk 1 above)
Comment (jwodder) [reply .github/workflows/template/template.yml.j2:1]: The script already determines the workflow names by templating `build-{{ostype}}.yml`.  I don't see an advantage from naming the template to match that.

Comment (yarikoptic) [reply .github/workflows/template/template.yml.j2:1]: To avoid looking for the right template if we get multiple in the future.

Comment (yarikoptic) [reply .github/workflows/template/template.yml.j2:1]: And if patches change

Comment (jwodder) [reply .github/workflows/template/template.yml.j2]: Done.

**Review by yarikoptic — CHANGES_REQUESTED** [review]
lets add testing for osx and even try to reenable for windows -- AFAIK that hanging issue was resolved

File: .github/workflows/template/template.yml.j2 (hunk 1 above)
Comment (yarikoptic) [reply .github/workflows/template/template.yml.j2]: I do not see any related changes, which commit?

Comment (jwodder) [reply .github/workflows/template/template.yml.j2]: This commit: https://github.com/datalad/git-annex/pull/22/commits/2a07b990b5858eac1a884a5ee378009210876189

Comment (yarikoptic) [reply .github/workflows/template/template.yml.j2]: ah, sorry, I conflated two issues in this single thread.  I was wondering about 

> And if patches change

i.e. adding `patches/*` into the list of paths for workflows to react on

Comment (jwodder) [reply .github/workflows/template/template.yml.j2]: The workflows now trigger on PRs that change `patches/*.patch`.

**Comment by jwodder** [comment]
@yarikoptic 
* The Datalad tests on Windows might be hanging, but only for the release version; it's still running after 2 hours, while all the other Datalad tests finished under 1h45m.
* The Datalad tests on macOS that require Docker are still failing due to being unable to download boot2docker, presumably because of GitHub API rate-limiting, despite passing a GitHub token to docker-machine.  The test run that doesn't require Docker, meanwhile, had several tests fail.

**Comment by yarikoptic** [comment]
> @yarikoptic Shouldn't this include `tmp/git-annex_*.dmg` for macOS?

addressed in 585951ebe3be73bb23543a56ed14a00fec8aedb7

**Comment by yarikoptic** [comment]
re failing tests on windows: filed https://github.com/datalad/datalad/issues/5210 -- we are getting close!!!

**Comment by yarikoptic** [comment]
I guess the plan would be to
- fixup osx build (WiP: https://github.com/datalad/git-annex/pull/28) first so we know that OSX builds is good when templated
- re-disable testing on OSX and windows where it fails (we gathered information and getting them to resolve, will start turning testing back up for maint/master as we think they are good enough)

**Comment by yarikoptic** [comment]
eh, OSX builds is still failing :-/  since no other outstanding PRs for workflows, I guess we could wait a bit more before proceeding here.

**Comment by yarikoptic** [comment]
To streamline path forward for #34 etc, and since we are in the "red" zone ATM anyways, should we may be just merge this PR as is?  or may be disable testing to the same choices as we have in master for now and then merge?

**Comment by jwodder** [comment]
@yarikoptic Yes, we should probably merge now.  I leave it up to you what tests to disable.

**Comment by yarikoptic** [comment]
ok, let's merge and aim for the bright future or later I would disable some runs

**Pull request merged**